src/VecSim/algorithms/hnsw/hnsw_factory.cpp

```diff
@@ -0,0 +1,75 @@
+#include "VecSim/algorithms/hnsw/hnsw_factory.h"
+#include "VecSim/algorithms/hnsw/hnsw_wrapper.h"
+#include "VecSim/algorithms/hnsw/hnswlib.h" //linklistsizeint
+#include "VecSim/vec_sim_common.h"          // labelType
+
+using namespace hnswlib;
+
+VecSimIndex *HNSWFactory::NewIndex(const HNSWParams *params,
+                                   std::shared_ptr<VecSimAllocator> allocator) {
+    // check if single and return new bf_index
+    assert(!params->multi);
+    return new (allocator) HNSWIndex(params, allocator);
+}
+
+size_t HNSWFactory::EstimateInitialSize(const HNSWParams *params) {
+
+    size_t est = sizeof(VecSimAllocator) + sizeof(size_t);
+    est += sizeof(HNSWIndex);
+    est += sizeof(hnswlib::HierarchicalNSW<float>) + sizeof(size_t);
+    // Used for synchronization only when parallel indexing / searching is enabled.
+#ifdef ENABLE_PARALLELIZATION
+    est += sizeof(VisitedNodesHandlerPool) + sizeof(size_t);
+#else
+    est += sizeof(VisitedNodesHandler) + sizeof(size_t);
+#endif
+    est += sizeof(tag_t) * params->initialCapacity + sizeof(size_t); // visited nodes
+
+    est += sizeof(void *) * params->initialCapacity + sizeof(size_t); // link lists (for levels > 0)
+    est += sizeof(size_t) * params->initialCapacity + sizeof(size_t); // element level
+    est += sizeof(size_t) * params->initialCapacity +
+           sizeof(size_t); // Labels lookup hash table buckets.
+
+    size_t size_links_level0 =
+        sizeof(linklistsizeint) + params->M * 2 * sizeof(idType) + sizeof(void *);
+    size_t size_total_data_per_element =
+        size_links_level0 + params->dim * VecSimType_sizeof(params->type) + sizeof(labelType);
+    est += params->initialCapacity * size_total_data_per_element + sizeof(size_t);
```

Comment:
Shouldn't this all be calculated only if `params->initialCapacity` is not 0?
Also, we can have `idType` and `labelType` here now.

Reply:
no, there is additional overhead regardless of the initial size